Canary: `MergeDeep` failures in `typescript@next` not detected by canary tests

I'm seeing a lot of these style of errors in the canary tests in my own modules:

```
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(144,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Destination>>>[number]' and 'UnknownArrayOrTuple'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(144,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Destination>>>[number]' and 'UnknownRecord'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(144,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Source>>>[number]' and 'UnknownArrayOrTuple'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(170,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Destination>>>[number]' and 'UnknownArrayOrTuple'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(170,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Destination>>>[number]' and 'UnknownRecord'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(170,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Source>>>[number]' and 'UnknownArrayOrTuple'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(196,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Destination>>>[number]' and 'UnknownArrayOrTuple'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(196,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Destination>>>[number]' and 'UnknownRecord'.
Error: ../node_modules/read-pkg/node_modules/type-fest/source/merge-deep.d.ts(196,5): error TS2321: Excessive stack depth comparing types 'PickRestType<ArrayTail<ArrayTail<Source>>>[number]' and 'UnknownArrayOrTuple'.
```

I didn't receive them in `typescript@5.4.0-dev.20231119` then started receiving them in `typescript@5.4.0-dev.20231121`.

To avoid a regression once `typescript@5.4.0` gets launched, we should track down the cause of this and file an issue with `typescript` if needed.

We should also track down why the canary tests in `type-fest` itself isn't catching it.

<!-- POLAR PLEDGE BADGE START -->
## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/sindresorhus) so you can upvote and help fund this issue.
- The funding will be given to active contributors.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/sindresorhus/type-fest/issues/784">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/sindresorhus/type-fest/issues/784/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/sindresorhus/type-fest/issues/784/pledge.svg">
</picture>
</a>
<!-- POLAR PLEDGE BADGE END -->

--- a/source/merge-deep.d.ts
+++ b/source/merge-deep.d.ts
@@ -63,7 +63,7 @@ type MergeDeepRecord<
 > = DoMergeDeepRecord<OmitIndexSignature<Destination>, OmitIndexSignature<Source>, Options>
 & Merge<PickIndexSignature<Destination>, PickIndexSignature<Source>>;
 
-// Utility to avoid compute ArrayTail twice.
+// Utility to avoid computing ArrayTail twice.
 type PickRestType___<Tail extends UnknownArrayOrTuple, Type> = Tail extends [] ? Type : PickRestType<Tail>;
 
 /**
@@ -82,7 +82,7 @@ type PickRestType<Type extends UnknownArrayOrTuple> = number extends Type['lengt
 	? PickRestType___<ArrayTail<Type>, Type>
 	: [];
 
-// Utility to avoid compute ArrayTail twice.
+// Utility to avoid computing ArrayTail twice.
 type OmitRestType___<Tail extends UnknownArrayOrTuple, Type extends UnknownArrayOrTuple, Result extends UnknownArrayOrTuple = []> = Tail extends [] ? Result : OmitRestType<Tail, [...Result, FirstArrayElement<Type>]>;
 
 /**
@@ -98,7 +98,6 @@ type Tuple5 = OmitRestType<[string, boolean[], ...number[]]>; // => [string, boo
 type Tuple6 = OmitRestType<string[]>; // => []
 ```
 */
-
 type OmitRestType<Type extends UnknownArrayOrTuple, Result extends UnknownArrayOrTuple = []> = number extends Type['length']
 	? OmitRestType___<ArrayTail<Type>, Type, Result>
 	: Type;

--- a/source/merge-deep.d.ts
+++ b/source/merge-deep.d.ts
@@ -63,8 +63,8 @@ type MergeDeepRecord<
 > = DoMergeDeepRecord<OmitIndexSignature<Destination>, OmitIndexSignature<Source>, Options>
 & Merge<PickIndexSignature<Destination>, PickIndexSignature<Source>>;
 
-// Utility to avoid computing ArrayTail twice.
-type PickRestType___<Tail extends UnknownArrayOrTuple, Type> = Tail extends [] ? Type : PickRestType<Tail>;
+// Helper to avoid computing ArrayTail twice.
+type PickRestTypeHelper<Tail extends UnknownArrayOrTuple, Type> = Tail extends [] ? Type : PickRestType<Tail>;
 
 /**
 Pick the rest type.
@@ -79,11 +79,17 @@ type Rest5 = PickRestType<string[]>; // => string[]
 ```
 */
 type PickRestType<Type extends UnknownArrayOrTuple> = number extends Type['length']
-	? PickRestType___<ArrayTail<Type>, Type>
+	? PickRestTypeHelper<ArrayTail<Type>, Type>
 	: [];
 
-// Utility to avoid computing ArrayTail twice.
-type OmitRestType___<Tail extends UnknownArrayOrTuple, Type extends UnknownArrayOrTuple, Result extends UnknownArrayOrTuple = []> = Tail extends [] ? Result : OmitRestType<Tail, [...Result, FirstArrayElement<Type>]>;
+// Helper to avoid computing ArrayTail twice.
+type OmitRestTypeHelper<
+    Tail extends UnknownArrayOrTuple,
+	Type extends UnknownArrayOrTuple,
+	Result extends UnknownArrayOrTuple = []
+> = Tail extends []
+    ? Result
+	: OmitRestType<Tail, [...Result, FirstArrayElement<Type>]>;
 
 /**
 Omit the rest type.
@@ -99,7 +105,7 @@ type Tuple6 = OmitRestType<string[]>; // => []
 ```
 */
 type OmitRestType<Type extends UnknownArrayOrTuple, Result extends UnknownArrayOrTuple = []> = number extends Type['length']
-	? OmitRestType___<ArrayTail<Type>, Type, Result>
+	? OmitRestTypeHelper<ArrayTail<Type>, Type, Result>
 	: Type;
 
 // Utility to avoid picking two times the type.

--- a/source/merge-deep.d.ts
+++ b/source/merge-deep.d.ts
@@ -84,11 +84,11 @@ type PickRestType<Type extends UnknownArrayOrTuple> = number extends Type['lengt
 
 // Helper to avoid computing ArrayTail twice.
 type OmitRestTypeHelper<
-    Tail extends UnknownArrayOrTuple,
+	Tail extends UnknownArrayOrTuple,
 	Type extends UnknownArrayOrTuple,
-	Result extends UnknownArrayOrTuple = []
+	Result extends UnknownArrayOrTuple = [],
 > = Tail extends []
-    ? Result
+	? Result
 	: OmitRestType<Tail, [...Result, FirstArrayElement<Type>]>;
 
 /**

--- a/source/merge-deep.d.ts
+++ b/source/merge-deep.d.ts
@@ -82,7 +82,6 @@ type PickRestType<Type extends UnknownArrayOrTuple> = number extends Type['lengt
 	? PickRestType___<ArrayTail<Type>, Type>
 	: [];
 
-
 // Utility to avoid compute ArrayTail twice.
 type OmitRestType___<Tail extends UnknownArrayOrTuple, Type extends UnknownArrayOrTuple, Result extends UnknownArrayOrTuple = []> = Tail extends [] ? Result : OmitRestType<Tail, [...Result, FirstArrayElement<Type>]>;
 

--- a/source/merge-deep.d.ts
+++ b/source/merge-deep.d.ts
@@ -63,6 +63,9 @@ type MergeDeepRecord<
 > = DoMergeDeepRecord<OmitIndexSignature<Destination>, OmitIndexSignature<Source>, Options>
 & Merge<PickIndexSignature<Destination>, PickIndexSignature<Source>>;
 
+// Utility to avoid compute ArrayTail twice.
+type PickRestType___<Tail extends UnknownArrayOrTuple, Type> = Tail extends [] ? Type : PickRestType<Tail>;
+
 /**
 Pick the rest type.
 
@@ -76,8 +79,12 @@ type Rest5 = PickRestType<string[]>; // => string[]
 ```
 */
 type PickRestType<Type extends UnknownArrayOrTuple> = number extends Type['length']
-	? ArrayTail<Type> extends [] ? Type : PickRestType<ArrayTail<Type>>
+	? PickRestType___<ArrayTail<Type>, Type>
 	: [];
+
+
+// Utility to avoid compute ArrayTail twice.
+type OmitRestType___<Tail extends UnknownArrayOrTuple, Type extends UnknownArrayOrTuple, Result extends UnknownArrayOrTuple = []> = Tail extends [] ? Result : OmitRestType<Tail, [...Result, FirstArrayElement<Type>]>;
 
 /**
 Omit the rest type.
@@ -92,8 +99,9 @@ type Tuple5 = OmitRestType<[string, boolean[], ...number[]]>; // => [string, boo
 type Tuple6 = OmitRestType<string[]>; // => []
 ```
 */
+
 type OmitRestType<Type extends UnknownArrayOrTuple, Result extends UnknownArrayOrTuple = []> = number extends Type['length']
-	? ArrayTail<Type> extends [] ? Result : OmitRestType<ArrayTail<Type>, [...Result, FirstArrayElement<Type>]>
+	? OmitRestType___<ArrayTail<Type>, Type, Result>
 	: Type;
 
 // Utility to avoid picking two times the type.

